semantics/src/namespace/types.rs

```diff
@@ -320,6 +380,12 @@ pub fn type_desc(
         fe::TypeDesc::Base { base: "bool" } => Ok(Type::Base(Base::Bool)),
         fe::TypeDesc::Base { base: "bytes" } => Ok(Type::Base(Base::Byte)),
         fe::TypeDesc::Base { base: "address" } => Ok(Type::Base(Base::Address)),
+        fe::TypeDesc::Base { base } if &base[..6] == "string" => {
+            let max_size = base[6..]
+                .parse::<usize>()
```

Comment:
I wonder if there is a reasonable upper bound that should be placed on the `max_size` other than `usize`. Especially because, if I understand correctly, `usize` will be determined by the machine that compiles the contract. So, in theory if someone uses `string18446744073709551616` (2**64) such contract will compile on a 64-bit machine but not on a 32-bit machine. So, maybe just parsing this into `u32` (or something even lower) would be better?

Reply:
I think a reasonable upper bound would be based on the how much memory you could allocate without surpassing the block gas limit. So I guess that would look something like this:

`3 * (size / 32) + ((size / 32)^2 / 256) = 10,000,000`

Plugging this into Wolfram, I get `size = 1,606,844`.

For comparison, `2^32` is `4,294,967,296`, so the limit based on gas is pretty far from that.

I think if we were to have a check in place, we should check the value against the number based on gas (whatever type it may be).

Reply:
reference: [memory fee on page 26](https://ethereum.github.io/yellowpaper/paper.pdf)

I've just replace `a` with `size / 32`, since that's how many words a string will occupy.

Reply:
Oh yes, I'm absolutely for choosing a more realistic upper bound. My main concern was that if we choose `usize` as upper bound then we all types that compile on one machine but doesn't on another. I mean, realistically, we can choose a number that is much much lower than `1,606,844`, too. How about we just pick a round number constant such as `1 000 000` or `100 000` or even just `10000`. Realistically, I can not imagine anything near those numbers to have practical use in the near term anyway.